runsc/container/container_test.go

```diff
@@ -3135,3 +3136,129 @@ func TestOverlayByMountAnnotation(t *testing.T) {
 		t.Fatalf("overlay filestore at %q was not deleted after container.Destroy()", filestoreFile)
 	}
 }
+
+// TestMountEROFS checks that the checksums from the target directory in container
+// are identical with the ones from the source directory on host.
+func TestMountEROFS(t *testing.T) {
+	// Create a temporary directory with some random files in it, which will
+	// be used as the source directory to create the EROFS images.
+	sourceDir, err := ioutil.TempDir(testutil.TmpDir(), "erofs-source")
+	if err != nil {
+		t.Fatalf("ioutil.TempDir() failed: %v", err)
+	}
+	testApp, err := testutil.FindFile("test/cmd/test_app/test_app")
+	if err != nil {
+		t.Fatalf("error finding test_app: %v", err)
+	}
+
+	// Source directory is a small directory. Let's create a big directory in it.
+	cmd := fmt.Sprintf("%s fsTreeCreate --target-dir=%s --create-symlink --depth=1 --file-per-level=500 --file-size=5000", testApp, filepath.Join(sourceDir, "big-directory"))
+	if out, err := exec.Command("/bin/sh", "-c", cmd).CombinedOutput(); err != nil {
+		t.Fatalf("exec: sh -c %q, err: %v, out: %s", cmd, err, out)
+	}
+
+	// Create a test script which can be used to get the checksums
+	// from a specified directory.
+	scriptFile := path.Join(sourceDir, "test-script")
+	if err := os.WriteFile(scriptFile, []byte(`#!/bin/bash
+set -u -e -o pipefail
+dir=$1
+find $dir -printf "%P\n" | sort | md5sum
+find $dir -type l | sort | xargs readlink | md5sum
```

Comment:
readlink(1) does not accept input from stdin. You need to do `xargs -I {} readlink {}`

Reply:
Oops, will fix it.. Thanks!

Reply:
In `xargs readlink`, `xargs(1)` will convert the lines from stdin to `readlink(1)`'s arguments. I did a quick test in ubuntu 23.04, it works:

```
$ ln -s 123 abc && ln -s foo bar && ls | xargs readlink
123
foo
```

Reply:
Hmm yeah actually this seems to be working for me too on my host. But in our test env, this test was failing with:
```
   third_party/gvisor/runsc/container/container_test.go:3176: exec: /build/work/f8fcd959590df7e8bf657328f9f0df8ecaeb/google3/tmp/erofs-source2064802333/test-script /build/work/f8fcd959590df7e8bf657328f9f0df8ecaeb/google3/tmp/erofs-source2064802333, err: exit status 123, out: 9a793baa190d9cc79644d7e39ec5ff80  -
        readlink: extra operand `/build/work/f8fcd959590df7e8bf657328f9f0df8ecaeb/google3/tmp/erofs-source2064802333/big-directory/file10'
        Try `readlink --help' for more information.
        d41d8cd98f00b204e9800998ecf8427e  -
```

Reply:
It worked with `xargs -I {} readlink {}`.

readlink(1) not receiving stdin input is discussed here too: https://unix.stackexchange.com/q/93334. Maybe is OS dependent.

Reply:
It seems in your test env, `readlink(1)` cannot accept multiple files in arguments.

The difference between `xargs readlink` and `xargs -I {} readlink {}` is that, in `xargs -I {} readlink {}`, `readlink(1)` will be called by `xargs(1)` multiple times and files will be passed to `readlink(1)` one by one as it implies `-L 1`. I will use what you suggested. Thanks!

Reply:
That makes sense. Actually `-L 1` is more readable as it conveys the intent more clearly. Could you switch to that?

Reply:
Sure, will do.